File: machine/cpp/machine/runtime/host_fault.cpp
#include "host_fault.h"

#include "machine/bus/io.h"
#include "machine/runtime/runtime.h"

namespace bmsx {

HostFaultState::HostFaultState(Runtime& runtime)
	: m_runtime(runtime) {
}

void HostFaultState::publishStartup() {
	m_runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_FLAGS, valueNumber(static_cast<double>(HOST_FAULT_FLAG_ACTIVE | HOST_FAULT_FLAG_STARTUP_BLOCKING)));
	m_runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_STAGE, valueNumber(static_cast<double>(HOST_FAULT_STAGE_STARTUP_AUDIO_REFRESH)));
}

void HostFaultState::clear() {
	m_runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_FLAGS, valueNumber(0.0));
	m_runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_STAGE, valueNumber(static_cast<double>(HOST_FAULT_STAGE_NONE)));
}

}

File: machine/ts/machine/runtime/host_fault.ts
import { convertToError } from '../../lua/value';
import {
	HOST_FAULT_FLAG_ACTIVE,
	HOST_FAULT_FLAG_STARTUP_BLOCKING,
	HOST_FAULT_STAGE_NONE,
	HOST_FAULT_STAGE_STARTUP_AUDIO_REFRESH,
	IO_SYS_HOST_FAULT_FLAGS,
	IO_SYS_HOST_FAULT_STAGE,
} from '../bus/io';
import { Runtime } from './runtime';

export class HostFaultState {
	constructor(private readonly runtime: Runtime) {
	}

	public publishStartup(error: unknown): void {
		const normalized = convertToError(error);
		const message = normalized.message.length > 0 ? normalized.message : String(error);
		const runtime = this.runtime;
		runtime.setGlobal('sys_host_fault_message', runtime.internString(message));
		runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_FLAGS, (HOST_FAULT_FLAG_ACTIVE | HOST_FAULT_FLAG_STARTUP_BLOCKING) >>> 0);
		runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_STAGE, HOST_FAULT_STAGE_STARTUP_AUDIO_REFRESH);
	}

	public clear(): void {
		const runtime = this.runtime;
		runtime.setGlobal('sys_host_fault_message', null);
		runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_FLAGS, 0);
		runtime.machine.memory.writeValue(IO_SYS_HOST_FAULT_STAGE, HOST_FAULT_STAGE_NONE);
	}

}
